### DISCUSSION TITLE
Proposal:  Multiple imports per using directive

### DISCUSSION THREAD

**Discussion by bondsbw** [dotnet/csharplang#1696] [Language Ideas]
Multiple namespace imports from a common base namespace are often used together:

```c#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using System.Xml;
```

This is a proposal to add a shorthand block syntax to the `using` directive in order to declare multiple imports at once:

 ```c#
using System;
using System.{Diagnostics, IO, Reflection, Text, Xml};

using System.Collections.Generic;
using System.Threading.Tasks;

using System.Windows;
using System.Windows.{Controls, Input, Media};
```

## Advantages
The cleaner syntax encourages grouping of similar namespaces together and reduces the line count.  Both factors improve the ability to review the namespace import list.

## Similar Proposal
#1421 provides a similar feature with similar syntax.  But according to comments, its nesting capability and usage of `this` keyword were deemed confusing.

The syntax from my proposal is based on other comments on that post (https://github.com/dotnet/csharplang/issues/1421#issuecomment-377343587 and https://github.com/dotnet/csharplang/issues/1421#issuecomment-377587174) which eliminate nesting and `this`.  These changes were viewed favorably based on comment votes.

I believe that creating a new proposal for these differences is warranted so that they do not get lost under a proposal which received many down votes.

**Comment by iam3yal** [comment]
@bondsbw The general idea is nice but tbh I really find this `using System.{Diagnostics, IO, Reflection, Text, Xml};` harder to read than the traditional way so I'd prefer something like this `using System.Diagnostics, IO, Reflection, Text, Xml;`

**Comment by iam3yal** [comment]
The issue with this kind of proposals is that if you have this many using statements:
```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using System.Xml;
```
Then it might be a hint you need to refactor and even then I'm not I'm sure it's worth it.

Just my 2c.

**Comment by bondsbw** [comment]
The braceless syntax is good too.  (Might need to think about whether #218 and other proposals for `using` would cause any conflicts... none that I can think of at the moment.)

**Comment by bondsbw** [comment]
Similar code is more likely to use similar imports.  More likely to see this after a good refactoring:

```c#
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using System.Xml;
```
than this:
```c#
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Controls;
using System.Xml;
```

So even though there are fewer imports, they would be the ones more likely to benefit from reduction.

**Comment by CyrusNajmabadi** [comment]
> The cleaner syntax encourages grouping of similar namespaces together

This is just a restatement of the proposal.   :)

> and reduces the line count. Both factors improve the ability to review the namespace import list.

TBH, this just doesn't seem to be that useful/important.

**Comment by CyrusNajmabadi** [comment]
I mean, just looking at your example:

> using System.{Diagnostics, IO, Reflection, Text, Xml};

These namespaces are wildly different.  WHy does having them all in one line make thigns easier to review than just seeing:

```
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Xml;
```

In both cases i can see you're using 5 wildly different namespaces... how does that help me?  How does consolidation make it any better?

--

Another issue is that i don't see how it helps me to know those 5 namespaces are used in one line... but then these two namespaces are called out in different line:

```
using System.Collections.Generic;
using System.Threading.Tasks;
```

Why is IO/Xml/Diagnostics really important to see together... but seeing Collections.Generic is not...

Basically... i'm not actually seeing how this makes things any clearer when i'm going to read a PR.

**Comment by bondsbw** [comment]
> This is just a restatement of the proposal. :)

What I meant was to contrast with

```c#
using System.Reflection;
using System.Text;
using System;
using System.Windows;
using System.Windows.Controls;
using MyStuff;
using MyStuff.CollectionsEx;
using Logger;
using System.Threading.Tasks;
using System.Windows.Input;
using TheirStuff.Net;
using System.Xml;
using System.Windows.Media;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
```

**Comment by bondsbw** [comment]
> Another issue is that i don't see how it helps me to know those 5 namespaces are used in one line... but then these two namespaces are called out in different line:
> ```
> using System.Collections.Generic;
> using System.Threading.Tasks;
> ```
> Why is IO/Xml/Diagnostics really important to see together... but seeing Collections.Generic is not...

That's fair.  I considered adding something like this to the proposal:

```c#
using System.{ Collections.Generic, Threading.Tasks };
```

**Comment by tannergooding** [comment]
```C#
using System.Reflection;
using System.Text;
using System;
using System.Windows;
using System.Windows.Controls;
using MyStuff;
using MyStuff.CollectionsEx;
using Logger;
using System.Threading.Tasks;
using System.Windows.Input;
using TheirStuff.Net;
using System.Xml;
using System.Windows.Media;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
```

This would be infinitely more readable if you ordered them appropriately:
```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;
using System.Xml;
using Logger;
using MyStuff;
using MyStuff.CollectionsEx;
using TheirStuff.Net;
```

**Comment by iam3yal** [comment]
@bondsbw 

> using System.Reflection;
> using System.Text;
> using System;
> using System.Windows;
> using System.Windows.Controls;
> using MyStuff;
> using MyStuff.CollectionsEx;
> using Logger;
> using System.Threading.Tasks;
> using System.Windows.Input;
> using TheirStuff.Net;
> using System.Xml;
> using System.Windows.Media;
> using System.Collections.Generic;
> using System.Diagnostics;
> using System.IO;

<s>How common it is to have that many namespaces in a single file?</s> and why aren't they sorted?

> That's fair. I considered adding something like this to the proposal

Don't you think that it's simple enough?

**Comment by tannergooding** [comment]
> How common it is to have that many namespaces in a single file

@eyalsk, it isn't that uncommon, especially when dealing with the UI stacks (XAML, WinForms, etc).

**Comment by iam3yal** [comment]
@tannergooding If you have that many namespaces trust me, you're doing it wrong.

**Comment by bondsbw** [comment]
Nah, I've seen that many times.  It happens quite often.

**Comment by tannergooding** [comment]
> If you have that many namespaces trust me, you're doing it wrong.

@eyalsk, no. It depends on your dependencies/etc. Even just going through Roslyn, there are plenty of files that have ~7 usings and quite a few that have over 12. Such as: http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.Xaml/Implementation/XamlTextViewCreationListener.cs,20e58dc1b7876041

When you are building UI components that depend on other layers of the stack, it is quite easy to have imports jump up.

I still am not a favor of this proposal, as I think having the IDE add/sort usings for you is sufficient; but having over 10 usings is not terribly unusual.

**Comment by CyrusNajmabadi** [comment]
> This would be infinitely more readable if you ordered them appropriately:

Agreed.  And note that this is something the IDE will do for you automatically now if you just 'format document':

![image](https://user-images.githubusercontent.com/4564579/42398865-6e6c39c4-8120-11e8-9766-68928d7330cd.png)

**Comment by tannergooding** [comment]
> And note that this is something the IDE will do for you automatically now if you just 'format document

This is exactly what I did to order them as well 😄. New project, paste usings, format document.

**Comment by iam3yal** [comment]
@bondsbw Whatever architectural pattern you use, let's say MVP and WinForms, your presenter shouldn't know about System.Windows.Forms.* and your logging infrastructure shouldn't be scattered all over the place... 

@tannergooding Okay, I'm not sure how many is "plenty" but I'll take that back, still think that if you have that many using statements it might be a hint for refactoring and in this particular case just by looking at the using statements it means that something does a little too much.

**Comment by CyrusNajmabadi** [comment]
Regardless of whether this happens a lot or not: the critical thing to understand is: what is the actual benefit here?

The substantive claim is that it: improve[s] the ability to review the namespace import list.

However, i'm not sure if that's actually true.  In all the examples i've seen so far i don't actually see anything about the 'single-line version' that is easier or clearer to me than the one-import-per-line approach.

So i'm not actually convinced this change is warranted, regardless of how common many-imports show up in practice.

**Comment by tannergooding** [comment]
> So i'm not actually convinced this change is warranted, regardless of how common many-imports show up in practice.

Right, this is my view as well. Even if a lot of namespaces are common practice, this falls into the same vein as putting multiple statements on a single line, which is generally viewed as "bad" since it generally decreases readability.

> that many using statements it might be a hint for refactoring and in this particular case just by looking at the using statements it means that something does a little too much.

@eyalsk, totally right. But there are also cases where you have a ~50 line file that has 8+ using statements because of things like attributes (MEF Export) and base classes (and things like needing `S.C.Immutable`): http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices.Xaml/CodeFixes/RemoveUnnecessaryUsings/XamlRemoveUnnecessaryUsingsCodeFixProvider.cs

**Comment by CyrusNajmabadi** [comment]
@tannergooding Right.  I'd much rather have something akin to VB's project-level imports.  After all, every IDE file has a *really* high chance of pulling in all of htese;

```
using System;
using System.Collections.Immutable;
using System.Composition;
using System.Threading;
using System.Threading.Tasks;
```

I'd prefer those 'core' system types just be always available**.  The file would then look like:

```
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.Editor.Xaml.Diagnostics;
using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
using Microsoft.CodeAnalysis.Shared.Extensions;
using Microsoft.VisualStudio.LanguageServices.Xaml;
using Roslyn.Utilities;
```

Which would really give me a sense of what's being used and how it's important for this type.  


--

** Obviously: i mean a project can opt-into deciding what is always available.  Not that those types are available in *all* projects implicitly.

**Comment by iam3yal** [comment]
@CyrusNajmabadi Yeah, definitely want something like that, we discussed [this](https://github.com/dotnet/roslyn/issues/14547) on the Roslyn repo.

**Comment by iam3yal** [comment]
@bondsbw What do you think about having something like @CyrusNajmabadi suggested? <s>will this solve part of your problem?</s>

**Comment by whoisj** [comment]
Perhaps allow for a `*` operator which means all of the namespaces starting with whatever.

example:
```C#
using System.*;
```

Then the 
![image](https://user-images.githubusercontent.com/11578325/42400196-a6f87a6e-813e-11e8-8d6d-947fb70dc5e5.png)
could be used to limit the set to the actual used namespaces. Just thinking out loud here.

**Comment by iam3yal** [comment]
@whoisj You will have to specify how deep you want to go though.

**Comment by bondsbw** [comment]
> will this solve part of your problem?

Who said it is my problem?  It provides an improvement in my opinion (folks are welcome to disagree), but it's not really a "problem".

I also like the idea of project level imports.

**Comment by bondsbw** [comment]
@whoisj 
> Perhaps allow for a * operator which means all of the namespaces starting with whatever.

Most languages use a wildcard to import all types within a namespace.  In C# that is the default behavior.

I'm not sure importing all namespaces under some namespace is particularly helpful.  (And I would think it could only hurt compiler performance.)

**Comment by iam3yal** [comment]
@bondsbw Poor wording on my part, sorry.

**Comment by whoisj** [comment]
> I'm not sure importing all namespaces under some namespace is particularly helpful. (And I would think it could only hurt compiler performance.)

Oh for sure - just throwing out ideas (as mentioned in the post)

**Comment by Mteheran** [comment]
if you don't want to see the namespace's list, just put them in a region

```
 #region namespaces

    using System;
    using System.Linq;
    using System.Web.Http;  
    using System.Threading;
    using System.Threading.Tasks;

 #endregion
```

**Comment by tannergooding** [comment]
@Mteheran, that requires an IDE (which already lets you collapse sequential using statements).

**Comment by bondsbw** [comment]
@Mteheran 
> if you don't want to see the namespace's list

That's not the point of the proposal.

**Comment by CyrusNajmabadi** [comment]
One of hte points was: ```and reduces the line count```.  it's reduced about as much as possible when it's only one line :D

**Comment by Mteheran** [comment]
in that case, I propose this syntax: 
```
using System[(Linq, Web, Xml)];
using System[(All)]; 

```

**Comment by CyrusNajmabadi** [comment]
As someone reading the code, how is that easier to understand than just having three separate imports?  What does this do to benefit me as a consumer/maintainer of this code?

**Comment by bondsbw** [comment]
@CyrusNajmabadi Mostly the same argument as object and collection initializers.  Grouping and DRY.

(This comment is about the original proposed syntax.  Not @Mteheran's.)

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Mostly the same argument as object and collection initializers. Grouping and DRY.

Those arguments don't hold those.  With Obj/Collection initializers what i get is the ability to know that i've created an actual instance at runtime that is fully initialized to that state i've provided.  This is especially valuable as previously that may have taken many statements, but i'm in a context that only allows an expression.  i.e.:

```c#
Foo(new Whatever { X = 1, Y = 2 ... });
```

The same does not apply to namespace/imports.  Imports do not represent an 'instance'.  There is no situations where i'm hampered by the need to do multiple imports over having a single one that would represent hte 'complete entity'.  

Those guys also let me express the idea of "i don't want anyone to be able to access this value until initialization is complete".  With statement-oriented initialization, there is nothing blocking me from accidentally leaking/using the variable prior to finishing this up.  With initializers, i can express that idea simply.  Again, no such concept holds for namespaces/imports.  They're simply a way to bring names into scope. There is no concept of "the names should not be in scope until i specify all the sub-namespaces i want to bring in", nor would there need to be :)

**Comment by bondsbw** [comment]
> This is especially valuable as previously that may have taken many statements, but i'm in a context that only allows an expression. i.e.:

Ha, I knew someone would bring that up.

Initializers are often used in the context of direct assignment.  They are a popular replacement for multiple assignment statements, regardless of their ability to be used in any expression context.  I doubt it's all about preventing leakage.

I really do think there is a readability improvement with such blocks.

**Comment by asyncawaitmvvm** [comment]
> The issue with this kind of proposals is that if you have this many using statements:
> 
> [...]
> 
> Then it might be a hint you need to refactor and even then I'm not I'm sure it's worth it.
> 
> Just my 2c.

This is a worthless "code smell".  The number of using statements has no correlation to what the code is doing except to say that complicated things must draw from many domains. Maybe you're one of those unfortunate guys who creates a separate source file for a 10-line enum and makes it impossible to keep a 100K project organized in your head.

**Comment by iam3yal** [comment]
@asyncawaitmvvm Emphasis on the word **might** but more importantly emphasis on your attitude.

**Comment by DavidArno** [comment]
@asyncawaitmvvm,

Ha, if you dislike folk then take up ten lines in a new file for an enum, [you'd really hate my code, as four lines is enough for me, and three of those are for the namespace block](https://github.com/DavidArno/SuccincT/blob/master/src/SuccincT/Unions/Variant.cs) 🤣

**Comment by yaakov-h** [comment]
@asyncawaitmvvm 

> The number of using statements has no correlation to what the code is doing except to say that complicated things must draw from many domains.

... which often triggers [CA1502](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1502-avoid-excessive-complexity).

I have yet to see a method that has low cyclomatic complexity whilst drawing from many domains, i.e. many `using`s.

**Comment by terrajobst** [comment]
@asyncawaitmvvm 

I appreciate the passionate discussion but I don't appreciate your tone towards others in the community ([one example](https://github.com/dotnet/csharplang/issues/1702#issuecomment-403665953), [second example](https://github.com/dotnet/csharplang/issues/1696#issuecomment-403666951)). This behavior is a violation of [our code of conduct](https://dotnetfoundation.org/code-of-conduct). If you keep doing this, we might have to block your account, which I really don't want to do. However, we need to ensure that we can have a productive discussion with our community and the technical aspects are challenging enough that we cannot tolerate abusive behavior.
